design/baremetal-operator/raid-disk-controller-naming.md

```diff
@@ -0,0 +1,213 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# raid-extension-for-physical-disks-and-controller
+
+## Status
+
+provisional
+
+## Summary
+
+This document provides the naming convention adopted to express
+``physical_disks`` and ``controller`` sections of a RAID configuration of
+a baremetal-host. The aim is to provide a uniform way of representation for
+hardware from various vendors, in a way that enhances operator experience.
+It also proposes an implementation for the same.
+
+## Motivation
+
+Vendors have different ways of defining the names for their physical disks
+and RAID controllers within Ironic. This leads to a complication for the
+operator where user experience is affected. There needs to be a uniform
+naming convention that abstracts this and increases convenience.
+
+### Goals
+
+The primary goal is to reach consensus on one naming convention to depict
+disk and RAID controller names in the baremetal-host ``hardware-raid``
+section.
```

Comment:
Is metal3 going to be the only part of the stack that understands the naming abstraction? So it would be *our* convention?

How does one normally identify the disks? Using values like we have in our `rootDeviceHints` structure?

Reply:
> Is metal3 going to be the only part of the stack that understands the naming abstraction? So it would be _our_ convention?

Yes. It will be our convention, that should be user friendly, and depicts vendor agnosticism. 

> How does one normally identify the disks? Using values like we have in our `rootDeviceHints` structure?

For RAID configurations, the physical disks are usually identified by various vendors as follows:

The iDRAC driver, with a conjured string like  ``Disk.Bay.0:Enclosure.Internal.0-1:RAID.Integrated.1-1``. Here's the [reference.](https://docs.openstack.org/ironic/latest/admin/drivers/idrac.html#raid-interface)

The iRMC driver uses a list of integers for disks. Here's the [reference.](https://docs.openstack.org/ironic/latest/admin/drivers/irmc.html)

Others follow their own conventions. The plan is to abstract these details into user friendly naming convention, since these are difficult and cumbersome for the user. The implementation will only focus on the iDRAC driver, but since all the vendors will implement the same names from a user interface perspective, I proposed a simple convention above.

Reply:
I phrased that question poorly.

What information does Ironic expect as identification for the disks?

Why isn't this abstraction in Ironic?

Reply:
It depends on the ``raid-interface`` in use, since the API call made to the corresponding BMC for the server has it's own way of accepting parameters. Ironic accepts those, and passes them on. So, for instance, the ``idrac-wsman`` interface accepts physical disks in an extended format like ``Disk.Bay.0:Enclosure.Internal.0-1:RAID.Integrated.1-1`` whereas the ``irmc`` interface for RAID accepts disks as simple integers like ``0``, ``1``, or ``2``.

As to why isn't the abstraction in Ironic, I'm not entirely sure. Since Ironic is operated by another operator, in a larger ecosystem I'd say the abstraction to use would be up to the operator.

Reply:
Is it possible to derive those vendor-specific IDs from values we have from inspection (even if it is data not in the host status fields today)? Could we reuse the rootDeviceHints structure to specify a device for consistency within our API then have a set of types that know how to do the conversion?

Reply:
The ``rootDeviceHints`` can be used only to mark the device/RAID volume to be used for root, since they are used by the ramdisk usually. And I'm unsure that the introspection data will provide us information required to create these vendor-specific IDs. 

In order to resolve the issue of generating correct vendor IDs, we can focus supporting most common use cases, where we can generate the correct vendor IDs, especially, say, in case of the ``idrac`` driver, where the string is complicated to predict.

I'll ask around the Ironic community to see if there is a way to get this data from introspection, or if we can patch the introspector for the purpose.

Reply:
You mention asking around within the Ironic community to see if there was a way to get the data.  What was answer?

Reply:
The approach I had in mind was this: metal3 could find the device names via Ironic, and then use that info to request a particular RAID and BIOS config. Could we populate the YAML using the extracted information?

Reply:
@dtantsur I agree. Opting for a vendor neutral solution would not get us the desired result. There will always be a need to translate this generic API component into a vendor-specific component. There's a proposal in ironic conductor to expose this ``physical disks`` and ``controller`` lists to be exposed as a vendor-passthru which can be used in metal3. 

This patch is necessary for upstream airship efforts where this is a legitimate use-case, that's why we're pursuing it. Really appreciate all your input :)

Reply:
> The approach I had in mind was this: metal3 could find the device names via Ironic

What to do when the device names are not available?

Reply:
Another idea: cannot we make the drivers in ironic support simply numbers in addition to full names? This would remove the need for new API on the ironic side and will make this feature available to everyone using ironic.

Reply:
> > The approach I had in mind was this: metal3 could find the device names via Ironic
> 
> What to do when the device names are not available?

The logic inside the BMO will make sure that the names be made available for the supported BMC types( for unsupported ones, we can simply throw an exception). And this will be done by ensuring disk names through ironic (whatever method should we chose to do so)

Reply:
> Another idea: cannot we make the drivers in ironic support simply numbers in addition to full names? This would remove the need for new API on the ironic side and will make this feature available to everyone using ironic.

That's a great idea. We should evaluate this further and break it down into an RFE that is implementable for the ironic conductor. @dtantsur What are you thinking?

Reply:
Yep. I'd merge this design with vendor-specific names (since we cannot guarantee the mapping for all vendors) and figure it out on the ironic side.